rust-src/id/src/identity_provider.rs

```diff
@@ -246,13 +246,462 @@ pub fn validate_request<P: Pairing, C: Curve<Scalar = P::ScalarField>>(
     }
 }
 
+pub fn validate_request<P: Pairing, C: Curve<Scalar = P::ScalarField>>(
+    pre_id_obj: &PreIdentityObject<P, C>,
+    context: IpContext<P, C>,
+) -> Result<(), Reason> {
+    let pub_info_for_ip = &pre_id_obj.pub_info_for_ip;
+    let id_cred_pub = &pre_id_obj.pub_info_for_ip.id_cred_pub;
+    let common_fields = &pre_id_obj.common_fields;
+    let poks_common = &pre_id_obj.poks.common_proof_fields;
+    let proof_acc_sk = &pre_id_obj.poks.proof_acc_sk;
+    // Verify signature:
+    let keys = &pub_info_for_ip.vk_acc.keys;
+    let threshold = pub_info_for_ip.vk_acc.threshold;
+
+    // message signed
+    let signed = Sha256::digest(&to_bytes(&pub_info_for_ip));
+
+    let reason = Reason::IncorrectProof; // TODO: introduce different reason
+
+    // Notice that here we provide all the verification keys, and the
+    // function `verify_accunt_ownership_proof` assumes that
+    // we have as many signatures as verification keys.
+    if !utils::verify_account_ownership_proof(&keys, threshold, &proof_acc_sk, signed.as_ref()) {
+        return Err(reason);
+    }
+
+    let mut transcript = RandomOracle::domain("PreIdentityProof");
+    let (verifier, witness) = validate_request_common(
+        &mut transcript,
+        id_cred_pub,
+        common_fields,
+        poks_common,
+        context,
+    )?;
+    let verifier_prf_regid = com_eq::ComEq {
+        commitment: common_fields.cmm_prf,
+        y:          context.global_context.on_chain_commitment_key.g,
+        g:          pub_info_for_ip.reg_id.clone(),
+        cmm_key:    verifier.first.second.cmm_key_1,
+    };
+    let prf_regid_witness = pre_id_obj.poks.prf_regid_proof.clone();
+
+    let verifier = verifier.add_prover(verifier_prf_regid);
+    let witness = AndWitness {
+        w1: witness,
+        w2: prf_regid_witness,
+    };
+    let proof = SigmaProof {
+        challenge: poks_common.challenge,
+        witness,
+    };
+
+    if verify(&mut transcript, &verifier, &proof) {
+        Ok(())
+    } else {
+        Err(Reason::IncorrectProof)
+    }
```

Comment:
It would be good to provide a little bit of high-level explanation for what is being done here. Essentially the things we discussed. In particular the backwards compatibility aspect should be mentioned, because otherwise it is hard to understand why we don't simplify the proofs further by reordering them to get more reuse.